c10/cuda/CUDAException.cpp

```diff
@@ -26,10 +109,192 @@ void c10_cuda_check_implementation(
   check_message.append("CUDA error: ");
   check_message.append(cudaGetErrorString(err));
   check_message.append(c10::cuda::get_cuda_check_suffix());
+  check_message.append("\n");
+  if (include_device_assertions) {
+    check_message.append(c10_retrieve_device_side_assertion_info());
+  }
 #endif
 
   TORCH_CHECK(false, check_message);
 }
 
+CUDAKernelLaunchRegistry::CUDAKernelLaunchRegistry()
+    : do_all_devices_support_managed_memory(
+          check_if_all_devices_support_managed_memory()) {
+  for (C10_UNUSED const auto _ : c10::irange(get_device_count())) {
+    uvm_assertions.emplace_back(nullptr, uvm_deleter);
+  }
+
+  kernel_launches.resize(max_kernel_launches);
+}
+
+uint32_t CUDAKernelLaunchRegistry::insert(
+    const std::string& launch_filename,
+    const std::string& launch_function,
+    const uint32_t launch_linenum,
+    const std::string& kernel_name,
+    const int32_t stream_id) {
+#ifdef TORCH_USE_CUDA_DSA
+  const std::lock_guard<std::mutex> lock(read_write_mutex);
+  const auto my_gen_number = generation_number++;
+  // TODO: It would probably be good to get a stack trace here so that
+  // we can better indicate which launch caused the failure.
+  kernel_launches[my_gen_number % max_kernel_launches] = {
+      launch_filename,
+      launch_function,
+      launch_linenum,
+      c10::get_backtrace(),
+      kernel_name,
+      get_device_id(),
+      stream_id,
+      my_gen_number};
+  return my_gen_number;
+#else
+  return 0;
+#endif
+}
+
+std::pair<std::vector<DeviceAssertionsData>, std::vector<CUDAKernelLaunchInfo>>
+CUDAKernelLaunchRegistry::snapshot() const {
+  // This is likely to be the longest-lasting hold on the mutex, but
+  // we only expect it to be called in cases where we're already failing
+  // and speed is no longer important
+  const std::lock_guard<std::mutex> lock(read_write_mutex);
+
+  std::vector<DeviceAssertionsData> device_assertions_data;
+  for (const auto& x : uvm_assertions) {
+    if (x) {
+      device_assertions_data.push_back(*x);
+    } else {
+      device_assertions_data.emplace_back();
+    }
+  }
+
+  return std::make_pair(device_assertions_data, kernel_launches);
+}
+
+DeviceAssertionsData* CUDAKernelLaunchRegistry::
+    get_uvm_assertions_ptr_for_current_device() {
+  // Need a lock here so there's not race-condition on creating the new device
+  // assertions buffer
+  const std::lock_guard<std::mutex> lock(read_write_mutex);
+
+  if (!do_all_devices_support_managed_memory) {
+    return nullptr;
+  }
+
+  const auto device_num = get_device_id();
+
+  // If we've already set up this GPU with managed memory, return a pointer to
+  // the managed memory
+  if (uvm_assertions.at(device_num)) {
+    std::cerr << "Creating a new uvm_assertions pointer" << std::endl;
```

Comment:
this is not a failure situation - why spam the output?

Reply:
I've removed this line - thanks!